### PR TITLE
Op 19004 ldap group attribute

### PR DESCRIPTION
**Pull request by Pranav-b-7**
We prefer small, well tested pull requests.

Please refer to [Contributing to Spinnaker](https://spinnaker.io/community/contributing/).

When filling out a pull request, please consider the following:

* Follow the commit message conventions [found here](https://spinnaker.io/community/contributing/submitting/).
* Provide a descriptive summary for your changes.
* If it fixes a bug or resolves a feature request, be sure to link to that issue.
* Add inline code comments to changes that might not be obvious.
* Squash your commits as you keep adding changes.
* Add a comment to @spinnaker/reviewers for review if your issue has been outstanding for more than 3 days.

Note that we are unlikely to accept pull requests that add features without prior discussion. The best way to propose a feature is to open an issue first and discuss your ideas there before implementing them.

https://devopsmx.atlassian.net/browse/OP-19004
https://devopsmx.atlassian.net/browse/OP-19005

### CODE REVIEW

File: gate-web/src/main/java/com/opsmx/spinnaker/gate/util/KeycloakAuthUtils.java [hunk 1]
```diff
@@ -71,17 +69,37 @@ public class KeycloakAuthUtils {
   @Value("${keycloak.token:}")
   private String token;
 
+  @Value("${security.oauth2.client.clientSecret:}")
+  private String clientSecret;
+
   public static final String USER_STORAGE_PROVIDER_TYPE =
       "org.keycloak.storage.UserStorageProvider";
 
+  private final String GROUPS = "groups";
```
Comment (rahul-chekuri) [inline gate-web/src/main/java/com/opsmx/spinnaker/gate/util/KeycloakAuthUtils.java:78]: make all these strings static

File: gate-web/src/main/java/com/opsmx/spinnaker/gate/util/KeycloakAuthUtils.java [hunk 2]
```diff
@@ -135,7 +159,9 @@ public void deleteLdapComponent() {
   }
 
   public ComponentRepresentation getLdapComponent() {
+    log.info("ldap name : {}", ldapName);
```
Comment (rahul-chekuri) [inline gate-web/src/main/java/com/opsmx/spinnaker/gate/util/KeycloakAuthUtils.java:162]: delete these logs

File: gate-web/src/main/java/com/opsmx/spinnaker/gate/util/KeycloakAuthUtils.java [hunk 3]
```diff
@@ -105,6 +123,12 @@ public void addLdapComponent(MultivaluedHashMap<String, String> config) {
     log.info("Successfully added ldap: {} ", componentRepresentation);
   }
 
+  public void addLdapComponent(ComponentRepresentation componentRepresentation) {
```
Comment (rahul-chekuri) [inline gate-web/src/main/java/com/opsmx/spinnaker/gate/util/KeycloakAuthUtils.java:126]: are we calling this from somewhere outside of this class? this cant be private?

File: gate-web/src/main/java/com/opsmx/spinnaker/gate/util/KeycloakAuthUtils.java [hunk 4]
```diff
@@ -226,6 +252,195 @@ public AuthProviderModel getDefaultModel(String name) {
...
+    componentRepresentation.setProviderId(GROUP_LDAP_MAPPER_PROVIDER);
+    return componentRepresentation;
+  }
+
+  @NotNull
+  private MultivaluedHashMap<String, String> populateLdapCompConfig(
+      AuthProviderModel authProviderModel) {
+    Map<String, String> ldapAuthConfig = authProviderModel.getConfig();
+    MultivaluedHashMap<String, String> ldapCompConfig = new MultivaluedHashMap<>();
+    if (ldapAuthConfig.containsKey("groupDN")) {
```
Comment (rahul-chekuri) [inline gate-web/src/main/java/com/opsmx/spinnaker/gate/util/KeycloakAuthUtils.java:396]: create a method that takes checks the key and adds it to the map. We can replace the below conditions with that method. This makes code elegant.

File: gate-web/src/main/java/com/opsmx/spinnaker/gate/service/LdapKeycloakAuthService.java [hunk 5]
```diff
@@ -49,13 +47,18 @@ public AuthProviderModel create(AuthProviderModel authProviderModel) {
     MultivaluedHashMap<String, String> config =
         hashMapToMultivaluedHashMap(authProviderModel.getConfig());
     keycloakAuthUtils.addLdapComponent(config);
+    keycloakAuthUtils.addLdapGroupMapper(authProviderModel);
```
Comment (rahul-chekuri) [inline gate-web/src/main/java/com/opsmx/spinnaker/gate/service/LdapKeycloakAuthService.java:50]: Don't we need to handle this for the update LDAP flow?

File: gate-web/src/main/java/com/opsmx/spinnaker/gate/service/SamlKeycloakAuthService.java [hunk 6]
```diff
@@ -49,6 +47,8 @@ public AuthProviderModel create(AuthProviderModel authProviderModel) {
     if (!Boolean.parseBoolean(enabled)) {
       keycloakAuthUtils.disableSamlIdentityProvider();
     }
+    keycloakAuthUtils.addIdpGroupMapper(authProviderModel);
+    keycloakAuthUtils.addUserAttributeProtocolMapper(authProviderModel);
```
Comment (rahul-chekuri) [inline gate-web/src/main/java/com/opsmx/spinnaker/gate/service/SamlKeycloakAuthService.java:51]: Don't we need to handle this at update SAML flow?

File: gate-web/src/main/java/com/opsmx/spinnaker/gate/util/KeycloakAuthUtils.java (hunk 1 above)
Comment (Pranav-b-7) [reply gate-web/src/main/java/com/opsmx/spinnaker/gate/util/KeycloakAuthUtils.java:78]: Making it static will load the data in the memory even before the object creation.. and morever i thing this class has to be feature flaged, so the data loads only during the object creation

File: docker_build/gate.yml [hunk 7]
```diff
@@ -134,27 +134,28 @@ keycloak:
...
-    syncRegistrations: "true"
-    allowKerberosAuthentication: "false"
-    useKerberosForPasswordAuthentication: "false"
-    cachePolicy: "DEFAULT"
-    usePasswordModifyExtendedOp: "false"
-    validatePasswordPolicy: "false"
-    trustEmail: "false"
-    changedSyncPeriod: "-1"
-    fullSyncPeriod: "-1"
+  ldapConfig:
```
[the remark this answers is not in the document]
Comment (Pranav-b-7) [reply docker_build/gate.yml]: Modified

File: gate-web/src/main/java/com/opsmx/spinnaker/gate/util/KeycloakAuthUtils.java (hunk 2 above)
Comment (Pranav-b-7) [reply gate-web/src/main/java/com/opsmx/spinnaker/gate/util/KeycloakAuthUtils.java]: Done

File: gate-web/src/main/java/com/opsmx/spinnaker/gate/util/KeycloakAuthUtils.java (hunk 3 above)
Comment (Pranav-b-7) [reply gate-web/src/main/java/com/opsmx/spinnaker/gate/util/KeycloakAuthUtils.java]: Done. Made it to private

File: gate-web/src/main/java/com/opsmx/spinnaker/gate/service/LdapKeycloakAuthService.java (hunk 5 above)
Comment (Pranav-b-7) [reply gate-web/src/main/java/com/opsmx/spinnaker/gate/service/LdapKeycloakAuthService.java:50]: Added the update flow

File: gate-web/src/main/java/com/opsmx/spinnaker/gate/service/SamlKeycloakAuthService.java (hunk 6 above)
Comment (Pranav-b-7) [reply gate-web/src/main/java/com/opsmx/spinnaker/gate/service/SamlKeycloakAuthService.java:51]: Added the update flow

File: gate-web/src/main/java/com/opsmx/spinnaker/gate/util/KeycloakAuthUtils.java (hunk 1 above)
Comment (rahul-chekuri) [reply gate-web/src/main/java/com/opsmx/spinnaker/gate/util/KeycloakAuthUtils.java:78]: once the feature flag is removed you are going to make it static ?

**Review by rahul-chekuri — APPROVED** [review]

**Pull request merged**